exist-core/src/test/java/org/exist/xquery/AbsolutePathTests.java

```diff
@@ -0,0 +1,160 @@
+    }
+
+    @Test
+    public void immediateLambdaContainsSlash() throws EXistException, PermissionDeniedException {
+        final String query = "(function() { /x })()";
+        assertXQStaticError(ErrorCodes.XPDY0002, 1, 15, compileQuery(query));
+    }
+
+    @Test
+    public void immediateLambdaContainsLoneSlash() throws EXistException, PermissionDeniedException {
+        final String query = "(function() { / })()";
+        assertXQStaticError(ErrorCodes.XPDY0002, 1, 15, compileQuery(query));
+    }
+
+    @Test
+    public void immediateLambdaContainsDoubleSlash() throws EXistException, PermissionDeniedException {
+        final String query = "(function() { //x })()";
+        assertXQStaticError(ErrorCodes.XPDY0002, 1, 15, compileQuery(query));
+    }
+
+    @Test
+    public void immediateLambdaWithDocumentAndDoubleSlash() throws EXistException, PermissionDeniedException {
+        final Source expected = elemSource("<result><x/><x/><x/></result>");
+
+        final String query = "let $f := function($d) { $d ! //x }\n" +
+                "let $d := document { <root><x/><x/><y><x/></y></root> }\n" +
+                "return\n" +
+                "  <result>{ $f($d) }</result>";
+        final Either<XPathException, Sequence> actual = executeQuery(query);
+
+        assertXQResultSimilar(expected, actual);
+    }
+
+    @Test
+    public void immediateLambdaWithDocumentAndSlash() throws EXistException, PermissionDeniedException {
+        final Source expected = elemSource("<root/>");
+
+        final String query = "let $f := function($d) { $d ! /root }\n" +
+                "let $d := document { <root/> }\n" +
+                "return\n" +
+                "  $f($d)";
+        final Either<XPathException, Sequence> actual = executeQuery(query);
+
+        assertXQResultSimilar(expected, actual);
+    }
+
+    @Test
+    public void immediateLambdaWithDocumentAndLoneSlash() throws EXistException, PermissionDeniedException {
+        final Source expected = docSource("<root/>");
+
+        final String query = "let $f := function($d) { $d ! / }\n" +
+                "let $d := document { <root/> }\n" +
+                "return\n" +
+                "  $f($d)";
+        final Either<XPathException, Sequence> actual = executeQuery(query);
+
+        assertXQResultSimilar(expected, actual);
+    }
+
+    @Test
+    public void topLevelAbsolutePath() throws EXistException, PermissionDeniedException {
+        final Sequence expected = new IntegerValue(1);
+
+        final String query = "count(//*)";
+        final Either<XPathException, Sequence> actual = executeQuery(query);
+
+        assertThatXQResult(actual, equalTo(expected));
+    }
+
+    /**
+     * Creates an Document6 Source form an XML String.
+     *
+     * @param str a string representation of XML.
+     *
+     * @return a Document Source.
+     */
+    private static Source docSource(final String str) {
```

Comment:
would this helper function be of value in either
- the abstract class XQueryCompilationTest
- the assertions in XQueryAssertions
- or a new class ~XMLDomMatchers or so

Reply:
I am asking this because I see myself searching for that code in the future and also would like to not have it copied in several test classes.

Reply:
I will move them back into `DiffMatcher` which is where they came from in FusionDB.